backend/analyze_tailcall_graph.ml

```diff
@@ -0,0 +1,419 @@
+        in
+        Edge.Hashset.add (successors t unknown) edge_from_unknown;
+        (* Initialize vertex's adjacency set *)
+        let edges_from_vertex = Edge.Hashset.create 10 in
+        Vertex.Tbl.replace t.adjacencies vertex edges_from_vertex;
+        vertex
+      | Some vertex -> vertex
+    in
+    vertex
+  ;;
+
+  let add_edge
+    t
+    ~(from : Vertex.t)
+    ~(to_ : Vertex.t)
+    ~(actual_position : [ `Tail | `Nontail ])
+    ~(original_position : Typedtree.position_and_tail_attribute)
+    =
+    let label =
+      let impossible_because ~case fmt =
+        Misc.fatal_errorf ("case " ^^ case ^^ " impossible because " ^^ fmt)
+      in
+      match original_position, actual_position with
+      (* CR less-tco: Clarify Unknown_position.
+
+         Right now we use it for "fake" applications (like tuplify) which might makes our
+         analysis less precise -- e.g. tuplify really is a leaf function, so it cannot form
+         a cycle in the call graph.
+      *)
+      | Unknown_position, `Tail -> `Unknown_tail
+      | Unknown_position, `Nontail -> `Unknown_nontail
+      | Not_tail_position Explicit_tail, _ ->
+        impossible_because
+          "[@tail] not allowed on applications not in tail position"
+          ~case:"Not_tail_position Explicit_tail, _"
+      | Tail_position Explicit_tail, `Nontail ->
+        impossible_because
+          "[@tail] was not optimized to a tailcall"
+          ~case:"Tail_position Explicit_tail, `Nontail"
+      | Tail_position Hint_tail, `Nontail ->
+        impossible_because
+          "[@tail hint] on application in tail position was not optimized to a tailcall"
+          ~case:"Tail_position Hint_tail, `Nontail"
+      | Tail_position Explicit_non_tail, `Tail
+      | Not_tail_position Explicit_non_tail, `Tail ->
+        impossible_because
+          "[@nontail] was optimized to a tailcall"
+          ~case:"_ Explicit_non_tail, `Tail"
+      | Tail_position (Explicit_tail | Hint_tail), `Tail -> `Requested_tail
+      | Tail_position Default_tail, `Tail -> `Inferred_tail
+      | Tail_position Explicit_non_tail, `Nontail -> `Requested_nontail
+      | Tail_position Default_tail, `Nontail -> `Inferred_nontail
+      | Not_tail_position Hint_tail, `Tail -> `Requested_tail
+      | Not_tail_position Default_tail, `Tail -> `Became_tail_after_optimizations
+      | Not_tail_position _, `Nontail -> `Not_in_tail_position
+    in
+    let label : Edge.Label.t =
+      match label with
+      | `Unknown_tail -> Explicit_tail_edge
+      | `Unknown_nontail -> Explicit_nontail_edge
+      | `Not_in_tail_position -> Explicit_nontail_edge
+      | `Requested_nontail -> Explicit_nontail_edge
+      | `Requested_tail -> Explicit_tail_edge
+      | `Became_tail_after_optimizations ->
+        (* This is a conservative approximation *)
+        Explicit_tail_edge
+      | `Inferred_tail -> Inferred_tail_edge
+      | `Inferred_nontail -> Inferred_nontail_edge
+    in
```

Comment:
I would produce the `Edge.Label.t` directly in the first match, and use comments instead of a polymorphic variant to explain anything that seems tricky.

Reply:
Done.